### PR TITLE
Optimize core and frontend

### PR DESCRIPTION
**Pull request by odersky**

### CODE REVIEW

**Comment by odersky** [comment]
test performance please

File: compiler/src/dotty/tools/dotc/typer/RefChecks.scala [hunk 1]
```diff
@@ -490,10 +491,9 @@ object RefChecks {
        */
       def missingTermSymbols: List[Symbol] =
         val buf = new mutable.ListBuffer[Symbol]
-        for bc <- clazz.baseClasses
-            sym <- bc.info.decls.toList
-            if sym.is(DeferredTerm) && !isImplemented(sym) && !ignoreDeferred(sym)
-        do buf += sym
+        for bc <- clazz.baseClasses; sym <- bc.info.decls.toList do
+          if sym.is(DeferredTerm) && !isImplemented(sym) && !ignoreDeferred(sym)
+            buf += sym
```
Comment (liufengyun) [inline compiler/src/dotty/tools/dotc/typer/RefChecks.scala:496]: `then` is required after #9859 -- the benchmarks fail due to this.

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Comment by odersky** [comment]
test performance please

**Review by liufengyun — APPROVED** [review]
LGTM

**Pull request merged**